### PR TITLE
feat(nnx): add DyT (Dynamic Tanh) normalization module

### LINKED ISSUE #5456: Adding Dynamic Tanh (DyT) element wise Operation (Paper from Meta)
### Summary

An important paper from Meta(FAIR), MIT, NYU and Princeton authors in late 2025 shows transformers can perform equally if not better using a Dynamic tanh operation instead of a Layer Normalization. The PR for this issue will aim to add this technique to the nnx src (norms) with minimal code. There is a successor paper to this as well featuring Dynamic erf (Derf) function for which i will open another issue.

I am attaching the related links here:
Paper: https://arxiv.org/pdf/2503.10622
Title: Transformers without Normalization
Code: https://github.com/jiachenzhu/DyT
Website: https://jiachenzhu.github.io/DyT/

cc: @cgarciae @vfdev-5 @samanklesaria 

### Additional context

Image

Image

### PR DESCRIPTION
**Pull request by Ashutosh0x**
## Summary

Implements the **DyT (Dynamic Tanh)** normalization-free layer as an `nnx.Module`, from [*Transformers without normalization*](https://arxiv.org/abs/2503.10622) (Zhu et al., CVPR 2025).

**Closes #5456**

## What is DyT?

DyT replaces statistical normalization (mean/variance reduction) with a learned pointwise bounding function:

```
y = γ · tanh(α · x) + β
```

| Parameter | Shape | Default Init | Role |
|-----------|-------|-------------|------|
| `alpha` | scalar | 0.5 | Input scaling before tanh |
| `scale` (γ) | `(num_features,)` | ones | Per-feature scale |
| `bias` (β) | `(num_features,)` | zeros | Per-feature bias |

### Key properties
- **Zero reduction operations** — no mean/variance computation across features
- **Eliminates cross-device allreduce** under tensor parallelism (vs. 4,224 allreduces/step for RMSNorm at 33 norm sites × 64 gradient accumulation steps)
- **Optimizer-robust** — no negative interaction with spectral optimizers like Muon (interaction: −0.10), unlike Derf (+0.66); see [Abouzeid 2026](https://arxiv.org/abs/2604.01563)
- **Drop-in replacement** — same `num_features` / `feature_axes` API as `LayerNorm` and `RMSNorm`

### Usage

```python
from flax import nnx

layer = nnx.DyT(num_features=768, rngs=nnx.Rngs(0))
y = layer(x)  # x: any shape with last dim = 768
```

## Changes

| File | Change |
|------|--------|
| `flax/nnx/nn/normalization.py` | Add `DyT` class (159 lines) |
| `flax/nnx/__init__.py` | Export `DyT` |
| `tests/nnx/nn/normalization_test.py` | Add `TestDyT` (13 test cases) |

### Test coverage
- Default initialization values
- Forward pass correctness (manual computation)
- Output shape preservation
- Output boundedness (tanh saturation)
- Gradient flow through all parameters (alpha, scale, bias)
- `float32` / `float16` dtype handling
- `use_scale=False`, `use_bias=False`, both disabled
- Custom `alpha_init`
- Multi-dimensional inputs (4D image-like tensors)
- JIT compatibility (`nnx.jit`)
- Zero-input identity

## Design decisions

1. **`alpha` as a learnable `Param`** (not a hyperparameter) — follows the paper's design where alpha adapts per-sublayer during training. The paper shows alpha self-corrects under different optimizers (e.g., 42% drop under Muon vs. 28% for Derf).

2. **No `mask` parameter** — unlike `LayerNorm`/`RMSNorm`, DyT has no reduction operations, so masking is not applicable. This keeps the API honest rather than accepting-and-ignoring a mask argument.

3. **Follows existing patterns** — parameter initialization, dtype promotion, feature_axes broadcasting all mirror `RMSNorm` / `LayerNorm` exactly.

### CODE REVIEW

**Comment by vfdev-5** [comment]
Closing as discussed here: https://github.com/google/flax/issues/5457#issuecomment-4499227717
